### PR TITLE
Allow topologySpreadConstraints configuration

### PR DESCRIPTION
**Pull request by justusbunsi**
## Description

This adds the ability to configure `topologySpreadConstraints` as addition to the existing option for `tolerations`.
It is an alternative to the `affinity` that is not properly capable of allowing multiple pods on one node but also
tries to spread them as much as possible across the cluster.

## Motivation and Context

The default replica value is 16 (right now) which can lead to a massive resource consumption on one node in smaller clusters. The idea for this addition is to allow users to specify how the pods (replicas) are being spread across the cluster. It gives more control over this Helm Release in smaller clusters where most worker nodes have taints.

As this Kubernetes feature exists since Kubernetes 1.19 and is only useful for a replica count > 1, this was taken into account.

## How to test this PR?

- Vary the usage of Kubernetes cluster version 1.18.x and later to see that it is skipped for older version than 1.19
- Mix different values for `.Values.replicas`, `.Values.mode` and `.Values.tolerations`  and use `.Values.topologySpreadConstraints` based on [official docs](https://kubernetes.io/docs/concepts/workloads/pods/pod-topology-spread-constraints/)

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation updated
- [ ] Unit tests added/updated

### CODE REVIEW

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

File: helm/minio/templates/statefulset.yaml [hunk 1]
```diff
@@ -1,6 +1,7 @@
 {{- if eq .Values.mode "distributed" }}
 {{ $poolCount := .Values.pools | int }}
 {{ $nodeCount := .Values.replicas | int }}
+{{ $calculatedReplicas := mul $poolCount $nodeCount }}
```
Comment (harshavardhana) [inline helm/minio/templates/statefulset.yaml]:
```suggestion
{{ $replicas := mul $poolCount $nodeCount }}
```

IT is visible that its calculated it is not necessary to be called as "calculated"

File: helm/minio/templates/statefulset.yaml [hunk 2]
```diff
@@ -56,7 +57,7 @@ spec:
     type: {{ .Values.StatefulSetUpdate.updateStrategy }}
   podManagementPolicy: "Parallel"
   serviceName: {{ template "minio.fullname" . }}-svc
-  replicas: {{ mul $poolCount $nodeCount }}
+  replicas: {{ $calculatedReplicas }}
```
Comment (harshavardhana) [inline helm/minio/templates/statefulset.yaml]:
```suggestion
  replicas: {{ $replicas }}
```

File: helm/minio/templates/statefulset.yaml [hunk 3]
```diff
@@ -167,6 +168,12 @@ spec:
     {{- with .Values.tolerations }}
       tolerations:
 {{ toYaml . | indent 8 }}
+    {{- end }}
+    {{- if and (gt $calculatedReplicas 1) (ge .Capabilities.KubeVersion.Major "1") (ge .Capabilities.KubeVersion.Minor "19") }}
```
Comment (harshavardhana) [inline helm/minio/templates/statefulset.yaml]:
```suggestion
    {{- if and (gt $replicas 1) (ge .Capabilities.KubeVersion.Major "1") (ge .Capabilities.KubeVersion.Minor "19") }}
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by justusbunsi** [comment]
@harshavardhana Thanks for your review and sorry for not responding earlier. I've applied your changes. Ready for another review.

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Review by harshavardhana — APPROVED** [review]

**Comment by justusbunsi** [comment]
Do I need to take any actions due to the failing build? ~Or is it unrelated? It seems that the logs are already cleaned up.~ Edit: The error seems unrelated, if I haven't missed something.

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Pull request merged**